feat: add foldertree

PhatJack opened the pull request:
### Summary

This PR introduces the new `FolderTree` component.

### Linked issue

#674 

### Changes

- Added `src/components/ui/folder-tree.tsx`
- Added demos at `src/components/demos/foldertree-badge.tsx`,
- Added docs at `src/content/docs/components/folder-tree.mdx`
- Updated registry and constants

### Checklist

- [x] Component added
- [x] Demo added
- [x] Registry updated
- [x] Documentation written

Adityakishore0 commented:
Cool, The current implementation works, but there are several areas that need to be addressed to improve usability, maintainability, and accessibility.

### 1. API Design
The API is too verbose and not developer-friendly. A cleaner, nested structure should be used for better readability and intuitive usage. The overall design should match modern component libraries, providing a consistent and scalable developer experience.
like look at these api structure 

```tsx
<FolderTree.Root defaultExpanded={['root']} defaultSelected="file1" onSelect={handleSelect}>
  <FolderTree.Item id="root" label="Root Folder">
    <FolderTree.Trigger />
    <FolderTree.Content>
      <FolderTree.Item id="folder1" label="Folder 1">
        <FolderTree.Trigger />
        <FolderTree.Content>
          <FolderTree.Item id="file1" label="File 1.txt" />
          <FolderTree.Item id="file2" label="File 2.txt" />
        </FolderTree.Content>
      </FolderTree.Item>
    </FolderTree.Content>
  </FolderTree.Item>
</FolderTree.Root>
```
---

### 2. Simplify Styling Props
There are too many props for styling such as heightClassName, widthClassName, and sizeClassName. Instead of exposing multiple styling props, rely on a single className prop. This gives developers full control to style the component using Tailwind utilities and avoids unnecessary API complexity.

---

### 3. Remove Redundant State Styling Props
Separate props like hoverClassName and selectedClassName are redundant since Tailwind already supports state-based styling through pseudo-classes and data attributes. Mixing these props with hardcoded defaults creates inconsistency and limits customization. All state styles should be handled through className to keep the API minimal and predictable.

---

### 4. Avoid Hardcoded Defaults
Some default styles and behaviors are mixed into the component and cannot be overridden. Defaults are acceptable, but they should never prevent developers from fully customizing appearance or behavior. The user’s provided className should always take precedence over internal defaults.

---

### 5. Performance and Memoization
The component may render many nested items, leading to unnecessary re-renders. Proper memoization should be applied to improve performance, especially when dealing with large or deeply nested folder structures.

---

### 6. Strong Type Safety
The current typing is not strict enough. Use clear and strongly typed interfaces to prevent misuse and provide a better developer experience. Strong typing ensures clarity, improves maintainability, and enhances editor tooling and autocompletion.

---

### 7. Accessibility and Keyboard Navigation
The implementation lacks accessibility features. Proper ARIA roles should be added. Support for aria-expanded and aria-selected is essential for screen readers. Full keyboard navigation should be implemented, including arrow key navigation, Enter, and Space key handling, to ensure the component is usable for all users.
or
leverage a library such as Radix UI to handle accessibility out of the box.

---

### Summary
- Simplify the API by reducing unnecessary props and relying on className for styling.
- Remove redundant hoverClassName and selectedClassName props in favor of Tailwind state styling.
- Avoid hardcoded styles and behaviors that block customization.
- Apply memoization to improve performance.
- Strengthen type safety with clear TypeScript definitions.
- Add comprehensive accessibility support, including ARIA roles and keyboard navigation.

Addressing these issues will make the component more flexible, performant, and accessible while improving the overall developer experience.

PhatJack commented:
Hey @Adityakishore0 
I have refactored based on what you recommended me to fix.

- refactored from single component with data props to compound components (`FolderTree.Root`, `FolderTree.Item`, `FolderTree.Content`, `FolderTree.Trigger`).
- implemented context-based state sharing between components for expandedIds, selectedId, and level tracking to avoid prop drilling.
- I have also added modified and untracked props with visual indicators (M/U badges) for version control status.
- automatic level-based indentation, and improved click handling with separate selection/expansion logic.
- add all necessary ARIA attributes like `role`, `aria-expanded`, `aria-selected`.

Review it again. Thanks 😊

PhatJack commented:
I have removed the `maxHeight` prop and changed to data attributes for better control for className.
About the `Context`, right now I can't find a better solution to replace it.
I think it works well and is optimized.

PhatJack commented:
If you have any solutions for the `Context`, please show me. I would be very grateful. 😊
